csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs

```diff
@@ -698,31 +696,94 @@ public IDisposableReadOnlyCollection<DisposableNamedOnnxValue> Run(IReadOnlyColl
         }
 
         /// <summary>
-        /// Ends profiling for the session. Returns the profile file name.
-        /// 
-        public string EndProfiling()
+        /// Create OrtIoBinding instance to bind pre-allocated buffers
+        /// to input/output
+        /// </summary>
+        /// <returns></returns>
+        public OrtIoBinding CreateIoBinding()
         {
-            IntPtr nameHandle = IntPtr.Zero;
-            string str = null;
+            return new OrtIoBinding(this);
+        }
 
-            IntPtr status = NativeMethods.OrtSessionEndProfiling(_nativeHandle,
-                                                                  NativeMemoryAllocator.DefaultInstance.Handle,
-                                                                  out nameHandle);
+        /// <summary>
+        /// This method runs inference on the OrtIoBinding instance
+        /// The method does not return anything. This is a lightweight version of 
+        /// RunWithBindingAndNames(). When you bind pre-allocated buffers to the output values
+        /// you may not want to fetch the outputs since you already have access to them so you can spare
+        /// the expense of fetching them and pairing with names.
+        /// You can still fetch the outputs by calling OrtIOBinding.GetOutputValues()
+        /// </summary>
+        /// <param name="runOptions"></param>
+        /// <param name="ioBinding"></param>
+        public void RunWithBinding(RunOptions runOptions, OrtIoBinding ioBinding)
+        {
+            NativeApiStatus.VerifySuccess(NativeMethods.OrtRunWithBinding(Handle, runOptions.Handle, ioBinding.Handle));
+        }
 
-            try
-            {
-                NativeApiStatus.VerifySuccess(status);
-                str = Marshal.PtrToStringAnsi(nameHandle);
-            }
-            finally
+        /// <summary>
+        ///  This method return a collection of DisposableNamedOnnxValue as in other interfaces
+        ///  Query names from OrtIoBinding object and pair then with the array of OrtValues returned
+        /// from OrtIoBinding.GetOutputValues()
+        /// 
+        /// </summary>
+        /// <param name="runOptions">RunOptions</param>
+        /// <param name="ioBinding">OrtIoBinding instance with bindings</param>
+        /// <param name="names">optional parameter. If you already know the names of the outputs you can save a native
+        /// call to retrieve output names. They will be paired with the returned OrtValues and combined into DisposbleNamedOnnxValues.
+        /// Otherwise, the method will retrieve output names from the OrtIoBinding instance.
+        /// It is an error if you supply a different number of names than the returned outputs</param>
+        public IDisposableReadOnlyCollection<DisposableNamedOnnxValue> RunWithBindingAndNames(RunOptions runOptions, OrtIoBinding ioBinding, string[] names = null)
```

Comment:
So this method performs the Run() and copies contents back to the host (for values bound to the device) to provide the DisposableNamedOnnxValue. Is this correct ?

Reply:
1) This method does not copy the content to the host. The outputs are stored in the OrtValues that were bound prior to Run() was called. So if the output was bound to a piece of pre-allocated memory, then it will be stored there. I can see the issue with returning DisposablenamesOnnxValues since the underlying memory may not be readable. Current use case does not require reading that memory but feed it right back to the input. So we simply rebind the value bound to output to input
2) Yes, the method is OrtIOBinding.GetOutputValues() and it is used here. However, if you bind a piece of pre-allocated memory you may not need to call that function since you still own that allocation and has access to it. We are still debating if we should provide read/write/copy access to native memory.

---
In reply to: [463244009](https://github.com/microsoft/onnxruntime/pull/4646#discussion_r463244009) [](ancestors = 463244009)